# Dataset: Austin Animal Center Shelter Outcomes

##### Contexto
El Centro de Animales de Austin es el refugio de animales sin sacrificio más grande de los Estados Unidos, que brinda atención y refugio a más de 18.000 animales cada año y participa en una variedad de iniciativas a nivel de condado, ciudad y estado para la protección y el cuidado de animales abandonados, en riesgo y entregados.

In [ ]:
# importación de librerias
import pandas as pd
import numpy as np

# configuración de pandas, para que no muestre número en notación científica
pd.set_option("display.float_format", lambda x: "%.2f" % x)

In [2]:
# carga de datos
df = pd.read_csv("./dataset/aac_shelter_outcomes.csv")

In [3]:
# mostrar primeras filas, para entender la estructura del dataset
df.head()

,age_upon_outcome,animal_id,animal_type,breed,color,date_of_birth,datetime,monthyear,name,outcome_subtype,outcome_type,sex_upon_outcome
0,2 weeks,A684346,Cat,Domestic Shorthair Mix,Orange Tabby,2014-07-07T00:00:00,2014-07-22T16:04:00,2014-07-22T16:04:00,NaN,Partner,Transfer,Intact Male
1,1 year,A666430,Dog,Beagle Mix,White/Brown,2012-11-06T00:00:00,2013-11-07T11:47:00,2013-11-07T11:47:00,Lucy,Partner,Transfer,Spayed Female
2,1 year,A675708,Dog,Pit Bull,Blue/White,2013-03-31T00:00:00,2014-06-03T14:20:00,2014-06-03T14:20:00,*Johnny,NaN,Adoption,Neutered Male
3,9 years,A680386,Dog,Miniature Schnauzer Mix,White,2005-06-02T00:00:00,2014-06-15T15:50:00,2014-06-15T15:50:00,Monday,Partner,Transfer,Neutered Male
4,5 months,A683115,Other,Bat Mix,Brown,2014-01-07T00:00:00,2014-07-07T14:04:00,2014-07-07T14:04:00,NaN,Rabies Risk,Euthanasia,Unknown


In [4]:
# obtener información general del dataset(dataframe)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78256 entries, 0 to 78255
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   age_upon_outcome  78248 non-null  object
 1   animal_id         78256 non-null  object
 2   animal_type       78256 non-null  object
 3   breed             78256 non-null  object
 4   color             78256 non-null  object
 5   date_of_birth     78256 non-null  object
 6   datetime          78256 non-null  object
 7   monthyear         78256 non-null  object
 8   name              54370 non-null  object
 9   outcome_subtype   35963 non-null  object
 10  outcome_type      78244 non-null  object
 11  sex_upon_outcome  78254 non-null  object
dtypes: object(12)
memory usage: 7.2+ MB


##### El cotenido de cada columna es entonces: 

0.   age_upon_outcome: Edad del animal en el momento de su salida del refugio.     
1.   animal_id: Identificador de cada animal.            
2.   animal_type: Gato, perro u otro (¡incluyendo al menos un murciélago!).          
3.   breed: Raza del animal. Muchos animales son mestizos genéricos, por ejemplo, "mestizo de pelo largo".                
4.   color: Color del pelaje del animal, si lo tiene.                
5.   date_of_birth: Fecha de nacimiento del animal.       
6.   datetime: Marca de tiempo del resultado.             
7.   monthyear: Mes y año por separado            
8.   name: Nombre del animal                  
9.   outcome_subtype: Subtipo del resultado       
10.  outcome_type: Destino final de este animal. Las opciones posibles incluyen traslado, eutanasia (por compasión) o adopción.         
11.  sex_upon_outcome: Sexo del animal

In [ ]:
# traducir las columnas del dataframe a español

# se crea un diccionario con los nombres originales y sus traducciones
columnas_espanol = {
    "age_upon_outcome": "edad_al_resultado",
    "animal_id": "id_animal",
    "animal_type": "tipo_animal",
    "breed": "raza",
    "color": "color",
    "date_of_birth": "fecha_nacimiento",
    "datetime": "fecha_hora",
    "monthyear": "mes_ano",
    "name": "nombre",
    "outcome_subtype": "subtipo_resultado",
    "outcome_type": "tipo_resultado",
    "sex_upon_outcome": "sexo_al_resultado",
}

# se aplican los nuevos nombres al dataframe (las traducciones)
df = df.rename(columns=columnas_espanol)

In [10]:
# verificación de que los nombres nuevos están correctos
print("--- Columnas renombradas ---")
print(df.columns)

--- Columnas renombradas ---
Index(['edad_al_resultado', 'id_animal', 'tipo_animal', 'raza', 'color',
       'fecha_nacimiento', 'fecha_hora', 'mes_ano', 'nombre',
       'subtipo_resultado', 'tipo_resultado', 'sexo_al_resultado'],
      dtype='object')


In [ ]:
# ahora se convierten las columnas de fecha

# primero se convierte la columna 'fecha_hora' a tipo datetime
df["fecha_hora"] = pd.to_datetime(df["fecha_hora"])

# se continua con la columna 'fecha_nacimiento'
df["fecha_nacimiento"] = pd.to_datetime(df["fecha_nacimiento"])

In [12]:
# verificar que el cambio se haya realizado correctamente
print("\n--- Tipos de datos actualizados ---")
df.info()


--- Tipos de datos actualizados ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78256 entries, 0 to 78255
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   edad_al_resultado  78248 non-null  object        
 1   id_animal          78256 non-null  object        
 2   tipo_animal        78256 non-null  object        
 3   raza               78256 non-null  object        
 4   color              78256 non-null  object        
 5   fecha_nacimiento   78256 non-null  datetime64[ns]
 6   fecha_hora         78256 non-null  datetime64[ns]
 7   mes_ano            78256 non-null  object        
 8   nombre             54370 non-null  object        
 9   subtipo_resultado  35963 non-null  object        
 10  tipo_resultado     78244 non-null  object        
 11  sexo_al_resultado  78254 non-null  object        
dtypes: datetime64[ns](2), object(10)
memory usage: 7.2+ MB


In [ ]:
# ahora se tiene que ver si existen valores únicos en ciertas columnas

# ¿Qué tipos de animales hay?
print("--- Tipos de Animales ---")
print(df["tipo_animal"].unique())

# ¿Qué tipos de resultados hay? (adopción, eutanasia, etc. - en general a que se refiere el resultado)
print("\n--- Tipos de Resultado ---")
print(df["tipo_resultado"].unique())

# ¿Cómo se ve la columna de edad?
print("\n--- Edades ---")
print(df["edad_al_resultado"].unique()[:20])  # mostrar solo los primeros 20 ejemplos

--- Tipos de Animales ---
['Cat' 'Dog' 'Other' 'Bird' 'Livestock']

--- Tipos de Resultado ---
['Transfer' 'Adoption' 'Euthanasia' 'Return to Owner' 'Died' 'Disposal'
 'Relocate' 'Missing' nan 'Rto-Adopt']

--- Edades ---
['2 weeks' '1 year' '9 years' '5 months' '4 months' '3 years' '1 month'
 '3 months' '2 years' '2 months' '4 years' '8 years' '3 weeks' '8 months'
 '12 years' '7 years' '5 years' '6 years' '5 days' '10 months']


**edad_al_resultado**, es la columna que requiere mas tratamiento
Las edades son texto. Entonces se tienen que convertir a un solo número para esto mejor se pasará todo a días.
También recordar que tiene 8 valores nulos (non-null 78248). Para los valores nulos se van a rellenar con "0 days" (asumiendo que son recién nacidos o desconocidos) esto por comodidad y para para que la función de conversión no falle.

In [16]:
# Tratamiento de edad_al_resultado

# 1. Se rellenan los 8 valores nulos
df["edad_al_resultado"] = df["edad_al_resultado"].fillna("0 days")

In [17]:
# 2. Se crea la función de conversión, es decir, de texto a días
def convertir_edad_a_dias(texto_edad):
    # texto_edad a veces puede ser un float, para evitar errores entonces se convierte a string
    texto_edad = str(texto_edad)

    if "year" in texto_edad or "years" in texto_edad:
        return int(texto_edad.split()[0]) * 365
    elif "month" in texto_edad or "months" in texto_edad:
        return int(texto_edad.split()[0]) * 30
    elif "week" in texto_edad or "weeks" in texto_edad:
        return int(texto_edad.split()[0]) * 7
    elif "day" in texto_edad or "days" in texto_edad:
        return int(texto_edad.split()[0])
    else:
        return 0  # esto porque puede ser 'Unknown' o formatos no reconocidos

In [18]:
# 3. Se llama a la función para crear la nueva columna 'edad_en_dias'
df["edad_en_dias"] = df["edad_al_resultado"].apply(convertir_edad_a_dias)

In [ ]:
# 4. Verificar el resultado (mostramos la columna vieja y la nueva)
print("\n--- Verificación de 'edad_en_dias' ---")
print(
    df[["edad_al_resultado", "edad_en_dias"]].head(10)
)  # mostrar las primeras 10 filas


--- Verificación de 'edad_en_dias' ---
  edad_al_resultado  edad_en_dias
0           2 weeks            14
1            1 year           365
2            1 year           365
3           9 years          3285
4          5 months           150
5          4 months           120
6            1 year           365
7           3 years          1095
8           1 month            30
9          3 months            90


Ahora se pasa al tratamiento de **sexo_al_resultado (sex_upon_outcome)**
Esta columna contiene dos datos: el sexo (Macho/Hembra) y si está esterilizado (Sí/No). También recordar que tiene 2 valores nulos. La intención es simplificar esta columna, para un mejor tratamiento de los datos, eso se realizará de la siguiente manera: 

In [ ]:
# 1. Primero se tiene que ver si existen y cuáles son los valores únicos
print("--- Valores Originales de 'sexo_al_resultado' ---")
print(df["sexo_al_resultado"].unique())

--- Valores Originales de 'sexo_al_resultado' ---
['Intact Male' 'Spayed Female' 'Neutered Male' 'Unknown' 'Intact Female'
 nan]


In [ ]:
# 2. Ahora se crea una función para extraer si está esterilizado/a
def es_esterilizado(texto):
    texto = str(texto)  # Convertimos a string para manejar los 'nan'
    if "Neutered" in texto or "Spayed" in texto:
        return "Si"
    elif "Intact" in texto:
        return "No"
    else:
        return "Desconocido"  # Esto es para 'Unknown' y los 'nan'

In [ ]:
# 3. Función para extraer el sexo
def obtener_sexo(texto):
    texto = str(texto)  # Para evitar errores se convierte a string
    if "Male" in texto:
        return "Macho"
    elif "Female" in texto:
        return "Hembra"
    else:
        return "Desconocido"  # Esto es para 'Unknown' y los 'nan'

In [ ]:
# 4. Ahora se divide "sexo_al_resultado" en las nuevas columnas aplicando las funciones
df["esterilizado"] = df["sexo_al_resultado"].apply(es_esterilizado)
df["sexo"] = df["sexo_al_resultado"].apply(obtener_sexo)

In [ ]:
# 5. Se verifica el resultado
print("\n--- Verificación de 'esterilizado' y 'sexo' ---")
print(df[["sexo_al_resultado", "esterilizado", "sexo"]].head(10))


--- Verificación de 'esterilizado' y 'sexo' ---
  sexo_al_resultado esterilizado         sexo
0       Intact Male           No        Macho
1     Spayed Female           Si       Hembra
2     Neutered Male           Si        Macho
3     Neutered Male           Si        Macho
4           Unknown  Desconocido  Desconocido
5       Intact Male           No        Macho
6           Unknown  Desconocido  Desconocido
7     Spayed Female           Si       Hembra
8     Intact Female           No       Hembra
9     Spayed Female           Si       Hembra


Se realizó el tratamiento para las columnas que más lo necesitaban, aún así se dejaron varios valores nulos anteriormente, ahora se manejarán estos valores nulos de la siguiente manera: 

In [ ]:
# se eliminan las filas con 'tipo_resultado' = nulo
print(f"Filas antes de eliminar nulos: {len(df)}")

# se eliminan las filas donde 'tipo_resultado' es nulo
df = df.dropna(subset=["tipo_resultado"])

print(f"Filas después de eliminar nulos: {len(df)}")

Filas antes de eliminar nulos: 78256
Filas después de eliminar nulos: 78244


**Seleccionar y guardar**
<br>
Ahora que ya se tienen datos más limpios y fácil de manipular, entonces se puede proceder a crear nuestro dataset final (df_limpio), esto solo contendrá las columnas que usaremos para el dashboard y el modelo de predicción. Entonces se seleccionan únicamente las columnas nuevas y limpias que creamos y las que se necesiten.

In [26]:
# 1. Se define entonces la lista de columnas que queremos conservar
columnas_finales = [
    'fecha_hora',       # Columna de fecha limpia
    'tipo_resultado',  
    'tipo_animal',
    'raza',
    'color',
    'edad_en_dias',   # Columna de edad numérica
    'esterilizado',   # Columna limpia
    'sexo'            # Columna limpia
]

In [27]:
# 2. Se crea el nuevo DataFrame 'df_limpio'
df_limpio = df[columnas_finales]

In [28]:
# 3. Para estar seguros de la información final de nuestro dataset limpio
print("\n--- Información del DataFrame Limpio Final ---")
df_limpio.info()


--- Información del DataFrame Limpio Final ---
<class 'pandas.core.frame.DataFrame'>
Index: 78244 entries, 0 to 78255
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   fecha_hora      78244 non-null  datetime64[ns]
 1   tipo_resultado  78244 non-null  object        
 2   tipo_animal     78244 non-null  object        
 3   raza            78244 non-null  object        
 4   color           78244 non-null  object        
 5   edad_en_dias    78244 non-null  int64         
 6   esterilizado    78244 non-null  object        
 7   sexo            78244 non-null  object        
dtypes: datetime64[ns](1), int64(1), object(6)
memory usage: 5.4+ MB


In [29]:
# 4. Se realiza una verificación más robusta, al mostrar las primeras filas del dataset limpio
print("\n--- Vista Previa del DataFrame Limpio ---")
print(df_limpio.head())


--- Vista Previa del DataFrame Limpio ---
           fecha_hora tipo_resultado tipo_animal                     raza  \
0 2014-07-22 16:04:00       Transfer         Cat   Domestic Shorthair Mix   
1 2013-11-07 11:47:00       Transfer         Dog               Beagle Mix   
2 2014-06-03 14:20:00       Adoption         Dog                 Pit Bull   
3 2014-06-15 15:50:00       Transfer         Dog  Miniature Schnauzer Mix   
4 2014-07-07 14:04:00     Euthanasia       Other                  Bat Mix   

          color  edad_en_dias esterilizado         sexo  
0  Orange Tabby            14           No        Macho  
1   White/Brown           365           Si       Hembra  
2    Blue/White           365           Si        Macho  
3         White          3285           Si        Macho  
4         Brown           150  Desconocido  Desconocido  


In [31]:
# Ahora con todo seleccionado y listo se puede guardar el archivo
# Se utiliza index=False para no guardar el índice de pandas
ruta_guardado = "./dataset"
df_limpio.to_csv(ruta_guardado + '/datos_limpios_refugio.csv', index=False)

print("\n'datos_limpios_refugio.csv' guardado")


'datos_limpios_refugio.csv' guardado
